File: site/scripts/build-index.ts
// Build everything the site serves from the repo, into site/public/ (gitignored):
//   data/index.json            catalog of plugins + skills (rendered by the Next.js pages)
//   downloads/<skill>.skill    zip of each skill folder, for installing one skill by hand
//   downloads/<plugin>.zip     zip of each whole plugin (plugin files at the zip root)
//   marketplace.json           the marketplace with every plugin as an `archive` source pointing at
//                              downloads/<plugin>.zip, so people without GitHub access can run
//                              `/plugin marketplace add https://<site>/marketplace.json`
//
// Usage:  node scripts/build-index.ts [--check] [--prev-catalog SRC] [--site-url URL]
//   --check             build into a temp dir and just report
//   --prev-catalog SRC  published catalog (path, URL, or "auto"; defaults to "auto" on Vercel).
//                       Skills whose content hash is unchanged keep its "updated" date, because
//                       Vercel's shallow clone can't give reliable git dates.
//   --site-url URL      absolute base for marketplace.json archive URLs. Defaults to
//                       metadata.site in production and the deployment's own URL on previews.
import { createHash } from "node:crypto";
import { mkdirSync, mkdtempSync, readFileSync, rmSync, statSync, writeFileSync, existsSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { zipSync, type Zippable } from "fflate";
import { REPO_URL, type Catalog, type Plugin, type Skill } from "../lib/catalog.ts";
import {
  contentHash, git, iterSkills, loadMarketplace, loadPluginManifest, loadPrevCatalog, type Marketplace,
  opt, pluginDir, rel, ROOT, walk,
} from "./lib.ts";

const SITE_PUBLIC = path.join(ROOT, "site", "public");
const ZIP_EPOCH = new Date(1980, 0, 2); // local time: zip stores local dates and rejects anything before 1980

const nowIso = () => new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00");

function lastModified(dir: string): string {
  return (git("log", "-1", "--format=%cI", "--", rel(dir)) ?? "").trim() || nowIso();
}

/** Zip dir to dest (entries under arcRoot/, or at the zip root if arcRoot is ""). Returns [bytes, sha256]. */
function zipDir(dir: string, dest: string, arcRoot: string): [number, string] {
  const files: Zippable = {};
  for (const f of walk(dir)) {
    const name = path.basename(f);
    if (name === ".gitkeep" || name === ".DS_Store" || f.split(path.sep).includes("__pycache__")) continue;
    const arc = path.relative(dir, f).split(path.sep).join("/");
    // keep the unix mode (e.g. executable scripts); os: 3 tells unzip to honour it
    files[arcRoot ? `${arcRoot}/${arc}` : arc] = [readFileSync(f), { attrs: (statSync(f).mode & 0xffff) << 16, os: 3 }];
  }
  const data = zipSync(files, { level: 9, mtime: ZIP_EPOCH });
  mkdirSync(path.dirname(dest), { recursive: true });
  writeFileSync(dest, data);
  return [data.byteLength, createHash("sha256").update(data).digest("hex")];
}

function firstParagraph(body: string): string {
  const para: string[] = [];
  for (const line of body.split(/\r?\n/)) {
    const s = line.trim();
    if (!s || s.startsWith("#") || s.startsWith("<!--")) {
      if (para.length) break;
      continue;
    }
    para.push(s);
  }
  return para.join(" ").slice(0, 400);
}

function siteUrl(mp: Marketplace): string {
  const override = opt("--site-url");
  if (override) return override.replace(/\/+$/, "");
  if (process.env.VERCEL_ENV && process.env.VERCEL_ENV !== "production" && process.env.VERCEL_URL)
    return `https://${process.env.VERCEL_URL}`;
  return (mp.metadata?.site ?? "").replace(/\/+$/, "");
}

export function build(out: string, prev: Catalog | null): Catalog {
  const mp = loadMarketplace();
  const dl = path.join(out, "downloads");
  mkdirSync(path.join(out, "data"), { recursive: true });
  mkdirSync(dl, { recursive: true });
  const prevSkills = new Map(prev?.plugins.flatMap((p) => p.skills.map((s) => [s.name, s] as const)) ?? []);
  const base = siteUrl(mp);

  const index: Catalog = {
    marketplace: {
      name: mp.name,
      description: String(mp.description ?? ""),
      version: mp.version ?? "",
      repo: REPO_URL,
      site: base,
      install: `/plugin marketplace add ${REPO_URL.replace("https://github.com/", "")}`,
      generated: nowIso(),
      commit: (git("rev-parse", "--short", "HEAD") ?? "").trim() || (process.env.VERCEL_GIT_COMMIT_SHA ?? "").slice(0, 7),
    },
    plugins: [],
  };
  const archiveEntries: Record<string, unknown>[] = [];

  for (const entry of mp.plugins) {
    const pdir = pluginDir(entry);
    const manifest = loadPluginManifest(pdir) ?? {};
    const skills: Skill[] = [];
    for (const sk of iterSkills(entry)) {
      if (sk.problems.length) continue;
      const [size] = zipDir(sk.dir, path.join(dl, `${sk.name}.skill`), sk.name);
      const digest = contentHash(sk.dir);
      const old = prevSkills.get(sk.name);
      const md = sk.metadata;
      skills.push({
        name: sk.name,
        description: sk.description,
        summary: firstParagraph(sk.body),
        body: sk.body.trim(),
        owner: String(md.owner ?? ""),
        status: (md.status ?? "draft") as Skill["status"],
        version: String(md.version ?? ""),
        connectors: md.connectors ?? [],
        has_scripts: existsSync(path.join(sk.dir, "scripts")),
        has_references: existsSync(path.join(sk.dir, "references")),
        updated: old?.content_hash === digest ? old.updated : lastModified(sk.dir),
        source: `${REPO_URL}/tree/main/${rel(sk.dir)}`,
        download: `downloads/${sk.name}.skill`,
        download_bytes: size,
        content_hash: digest,
      });
    }
    const [psize, sha256] = zipDir(pdir, path.join(dl, `${entry.name}.zip`), "");
    const author = (entry.author ?? manifest.author) as { name?: string } | undefined;
    const plugin: Plugin = {
      name: entry.name,
      displayName: String(entry.displayName ?? manifest.displayName ?? entry.name),
      description: String(entry.description ?? manifest.description ?? ""),
      category: String(entry.category ?? ""),
      tags: (entry.tags ?? manifest.keywords ?? []) as string[],
      version: String(entry.version ?? manifest.version ?? ""),
      author: author?.name ?? "",
      install: `/plugin install ${entry.name}@${mp.name}`,
      source: `${REPO_URL}/tree/main/${rel(pdir)}`,
      download: `downloads/${entry.name}.zip`,
      download_bytes: psize,
      content_hash: contentHash(pdir),
      skills,
    };
    index.plugins.push(plugin);
    archiveEntries.push({ ...entry, source: { source: "archive", url: `${base}/${plugin.download}`, sha256 } });
  }

  writeFileSync(path.join(out, "data", "index.json"), JSON.stringify(index, null, 2) + "\n");
  // Same marketplace (same name, so installs are interchangeable), minus git-only fields.
  const { metadata, ...rest } = mp;
  const { pluginRoot: _unused, ...meta } = metadata ?? {};
  writeFileSync(path.join(out, "marketplace.json"), JSON.stringify({ ...rest, metadata: meta, plugins: archiveEntries }, null, 2) + "\n");
  return index;
}

const check = process.argv.includes("--check");
const prev = await loadPrevCatalog(loadMarketplace());
let idx: Catalog;
if (check) {
  const td = mkdtempSync(path.join(tmpdir(), "catalog-"));
  idx = build(td, prev);
  rmSync(td, { recursive: true, force: true });
} else {
  // clean generated outputs so removed skills disappear
  for (const sub of ["data", "downloads", "marketplace.json"]) rmSync(path.join(SITE_PUBLIC, sub), { recursive: true, force: true });
  idx = build(SITE_PUBLIC, prev);
}
const nSkills = idx.plugins.reduce((a, p) => a + p.skills.length, 0);
console.log(`${check ? "checked" : "built"} index: ${idx.plugins.length} plugins, ${nSkills} skills` +
  (check ? "" : ` → ${rel(SITE_PUBLIC)}/data/index.json, marketplace.json`));

File: scripts/build_index.py
#!/usr/bin/env python3
"""Build the data the marketplace site reads.

Writes:
  site/data/index.json          catalog of plugins + skills
  site/downloads/<skill>.skill  zip of each skill folder (for people without GitHub)
  site/downloads/<plugin>.zip   zip of each whole plugin

Usage:  python3 scripts/build_index.py [--check]
  --check   build to a temp dir and just report; don't write into site/
"""
from __future__ import annotations

import json
import shutil
import subprocess
import sys
import tempfile
import zipfile
from datetime import datetime, timezone
from pathlib import Path

from common import ROOT, iter_skills, load_marketplace, load_plugin_manifest, plugin_dir

SITE = ROOT / "site"
REPO_URL = "https://github.com/wwtdigital/skills-marketplace"


def git(*args) -> str:
    try:
        return subprocess.check_output(["git", *args], cwd=ROOT, text=True, stderr=subprocess.DEVNULL).strip()
    except Exception:
        return ""


def last_modified(path: Path) -> str:
    iso = git("log", "-1", "--format=%cI", "--", str(path.relative_to(ROOT)))
    return iso or datetime.now(timezone.utc).isoformat(timespec="seconds")


def zip_dir(src: Path, dest: Path, arc_root: str) -> int:
    dest.parent.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(dest, "w", zipfile.ZIP_DEFLATED) as z:
        for f in sorted(src.rglob("*")):
            if f.is_file() and f.name != ".gitkeep" and "__pycache__" not in f.parts:
                z.write(f, f"{arc_root}/{f.relative_to(src)}")
    return dest.stat().st_size


def first_heading_para(body: str) -> str:
    """First non-heading paragraph of the body, for the card summary."""
    para = []
    for line in body.splitlines():
        s = line.strip()
        if not s or s.startswith("#") or s.startswith("<!--"):
            if para:
                break
            continue
        para.append(s)
    return " ".join(para)[:400]


def build(out: Path) -> dict:
    mp = load_marketplace()
    data_dir, dl_dir = out / "data", out / "downloads"
    data_dir.mkdir(parents=True, exist_ok=True)
    dl_dir.mkdir(parents=True, exist_ok=True)

    index = {
        "marketplace": {
            "name": mp["name"],
            "description": mp.get("description", ""),
            "version": mp.get("version", ""),
            "repo": REPO_URL,
            "install": f"/plugin marketplace add wwtdigital/skills-marketplace",
            "generated": datetime.now(timezone.utc).isoformat(timespec="seconds"),
            "commit": git("rev-parse", "--short", "HEAD"),
        },
        "plugins": [],
    }
    for entry in mp["plugins"]:
        pdir = plugin_dir(entry)
        manifest = load_plugin_manifest(pdir) or {}
        skills = []
        for sk in iter_skills(entry):
            if sk.problems:
                continue
            skill_zip = dl_dir / f"{sk.name}.skill"
            size = zip_dir(sk.path, skill_zip, sk.name)
            skills.append({
                "name": sk.name,
                "description": sk.description,
                "summary": first_heading_para(sk.body),
                "owner": sk.metadata.get("owner", ""),
                "status": sk.metadata.get("status", "draft"),
                "version": str(sk.metadata.get("version", "")),
                "connectors": sk.metadata.get("connectors") or [],
                "has_scripts": (sk.path / "scripts").exists(),
                "has_references": (sk.path / "references").exists(),
                "updated": last_modified(sk.path),
                "source": f"{REPO_URL}/tree/main/{sk.path.relative_to(ROOT)}",
                "download": f"downloads/{sk.name}.skill",
                "download_bytes": size,
            })
        plugin_zip = dl_dir / f"{entry['name']}.zip"
        psize = zip_dir(pdir, plugin_zip, entry["name"])
        index["plugins"].append({
            "name": entry["name"],
            "displayName": entry.get("displayName") or manifest.get("displayName") or entry["name"],
            "description": entry.get("description") or manifest.get("description", ""),
            "category": entry.get("category", ""),
            "tags": entry.get("tags") or manifest.get("keywords") or [],
            "version": entry.get("version") or manifest.get("version", ""),
            "author": (entry.get("author") or manifest.get("author") or {}).get("name", ""),
            "install": f"/plugin install {entry['name']}@{mp['name']}",
            "source": f"{REPO_URL}/tree/main/{pdir.relative_to(ROOT)}",
            "download": f"downloads/{entry['name']}.zip",
            "download_bytes": psize,
            "skills": skills,
        })
    (data_dir / "index.json").write_text(json.dumps(index, indent=2) + "\n")
    return index


def main() -> int:
    check = "--check" in sys.argv
    if check:
        with tempfile.TemporaryDirectory() as td:
            idx = build(Path(td))
    else:
        # clean generated outputs so removed skills disappear
        for sub in ("data", "downloads"):
            shutil.rmtree(SITE / sub, ignore_errors=True)
        idx = build(SITE)
    n_sk = sum(len(p["skills"]) for p in idx["plugins"])
    print(f"{'checked' if check else 'built'} index: {len(idx['plugins'])} plugins, {n_sk} skills"
          + ("" if check else f" → {SITE.relative_to(ROOT)}/data/index.json"))
    return 0


if __name__ == "__main__":
    sys.exit(main())
